workers/jobs/google_scraper_test.go

```diff
@@ -0,0 +1,207 @@
+package jobs
+
+import (
+	"errors"
+	"net/http"
+	"os"
+	"testing"
+
+	"github.com/bxcodec/faker/v3"
+	"github.com/dnaeon/go-vcr/recorder"
+	"github.com/gin-gonic/gin"
+	"github.com/gocraft/work"
+	"github.com/gutakk/go-google-scraper/config"
+	"github.com/gutakk/go-google-scraper/db"
+	"github.com/gutakk/go-google-scraper/models"
+	"github.com/gutakk/go-google-scraper/services/google_scraping_service"
+	testDB "github.com/gutakk/go-google-scraper/tests/db"
+	"github.com/gutakk/go-google-scraper/tests/path_test"
+	"github.com/stretchr/testify/suite"
+	"golang.org/x/crypto/bcrypt"
+	"gopkg.in/go-playground/assert.v1"
+	"gorm.io/driver/postgres"
+	"gorm.io/gorm"
+)
+
+func init() {
+	gin.SetMode(gin.TestMode)
+
+	if err := os.Chdir(path_test.GetRoot()); err != nil {
+		panic(err)
+	}
+
+	config.LoadEnv()
+}
+
+type KeywordScraperDBTestSuite struct {
+	suite.Suite
+	userID   uint
+	enqueuer *work.Enqueuer
+}
+
+func (s *KeywordScraperDBTestSuite) SetupTest() {
+	database, _ := gorm.Open(postgres.Open(testDB.ConstructTestDsn()), &gorm.Config{})
+	db.GetDB = func() *gorm.DB {
+		return database
+	}
+
+	db.GenerateRedisPool()
+
+	testDB.InitKeywordStatusEnum(db.GetDB())
+	_ = db.GetDB().AutoMigrate(&models.User{}, &models.Keyword{})
+
+	hashedPassword, _ := bcrypt.GenerateFromPassword([]byte(faker.Password()), bcrypt.DefaultCost)
+	user := models.User{Email: faker.Email(), Password: string(hashedPassword)}
+	db.GetDB().Create(&user)
+	s.userID = user.ID
+
+	s.enqueuer = work.NewEnqueuer("test-job", db.GetRedisPool())
+}
+
+func (s *KeywordScraperDBTestSuite) TearDownTest() {
+	db.GetDB().Exec("DELETE FROM keywords")
+	db.GetDB().Exec("DELETE FROM users")
+	_, _ = db.GetRedisPool().Get().Do("DEL", testDB.RedisKeyJobs("test-job", "scraping"))
+}
+
+func TestKeywordScraperDBTestSuite(t *testing.T) {
+	suite.Run(t, new(KeywordScraperDBTestSuite))
+}
+
+func (s *KeywordScraperDBTestSuite) TestPerformScrapingJobWithValidJob() {
+	r, _ := recorder.New("tests/fixture/vcr/valid_keyword")
+	requestFunc := google_scraping_service.Request
+	google_scraping_service.Request = func(keyword string, transport http.RoundTripper) (*http.Response, error) {
+		url := "https://www.google.com/search?q=AWS"
+		client := &http.Client{Transport: r}
+
+		req, _ := http.NewRequest("GET", url, nil)
+		req.Header.Set("User-Agent", "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/87.0.4280.88 Safari/537.36")
+
+		resp, err := client.Do(req)
+		if err != nil {
+			return nil, err
+		}
+
+		return resp, err
+	}
+	defer func() { google_scraping_service.Request = requestFunc }()
```

Comment:
I don't get how this mock works as it seems to reimplement the whole Request method 🤔

Reply:
In this case, in order to use VCR I need to set `Transport` in the client and the `Request` function is called in `PerformSearchJob` and it is executed by worker so I cannot find the way how can I pass value to set the `Transport` in this case. Thus I need to override the request function to set the `Transport`.

Reply:
But I think as this is a unit test. I can just mock to return valid response (not to be response from google) and just check that job err is not nil as the function return `error`

What do you think?

Reply:
Fixed in def6a406845fa78cb7ca74d506b105cb3ffaad8d

Reply:
Indeed, just mocking the return value would be enough. That's what a mock should do in a unit test. I checked the fix but I still don't fully understand how it actually works 😓 

In `SetupTest`, `requestFunc` and `parsingFunc` are set to teh actua; implementation for `google_search_service.Request` and `google_search_service.ParseGoogleResponse`. So no mock implementation here, right?

https://github.com/gutakk/go-google-scraper/blob/ff0ee1455035e6a5649032ae7656306ab58efd1c/workers/jobs/google_scraper_test.go#L55-L59

And then the mocks are implemented in `TestPerformSearchJobWithValidJob`:

https://github.com/gutakk/go-google-scraper/blob/ff0ee1455035e6a5649032ae7656306ab58efd1c/workers/jobs/google_scraper_test.go#L80-L86

So the part I am not sure is how and why using `defer` 🤔

Reply:
Actually, after some googling around, `defer` is used to restore back to the original/application implementation. I understand that `defer` part now. 

But then why doing:

```
s.requestFunc = google_search_service.Request 
s.parsingFunc = google_search_service.ParseGoogleResponse 
```

Only `TestPerformSearchJobWithValidJob` implements the mocks while the other tests like `TestPerformSearchJobWithoutKeywordID` do not or implement only one 🤔 It is kind of unexpected and could lead to hard-to-see and hard-to-solve issues. To be truly a unit test (so not an integration test), all test should implement the mocks. Or the setup method should implement a general-purpose mock.

Reply:
Also, as a recommendation, it would make `SetupTest` more readable by  externalizing to methods"

```go
func (s *KeywordScraperDBTestSuite) SetupTest() {
  // these methods could be shared to a test helpers module
  database, _ := testHelpers.SetiupDatabase()
  s.userID = testHelpers.createUser()
  
  // This can be a private method in this test file/package
  setupMocks()

  s.enqueuer = work.NewEnqueuer("test-job", db.GetRedisPool())
}
```

It would make tests easier to manage and maintain.

Reply:
Added setupMocks() in aeca2ef754d0dc4cbee5ae1ece1b3837bd432b05.
For another setup I will do it in the test refactor issue.